storage/src/main/java/tech/pegasys/artemis/storage/ChainStorageClient.java

```diff
@@ -306,4 +306,46 @@ public void onNewSlot(SlotEvent slotEvent) {
       }
     }
   }
+
+  public Optional<BeaconBlock> getBlockBySlot(final UnsignedLong slot) {
+    return getBlockRootBySlot(slot).map(blockRoot -> store.getBlock(blockRoot));
+  }
+
+  public boolean isIncludedInBestState(final Bytes32 blockRoot) {
+    if (store == null) {
+      return false;
+    }
+    final BeaconBlock block = store.getBlock(blockRoot);
+    if (block == null) {
+      return false;
+    }
+    return getBlockRootBySlot(block.getSlot())
+        .map(actualRoot -> actualRoot.equals(block.hash_tree_root()))
+        .orElse(false);
+  }
+
+  private Optional<Bytes32> getBlockRootBySlot(final UnsignedLong slot) {
+    if (store == null || Bytes32.ZERO.equals(bestBlockRoot)) {
+      return Optional.empty();
+    }
+    final UnsignedLong slotsPerHistoricalRoot =
+        UnsignedLong.valueOf(Constants.SLOTS_PER_HISTORICAL_ROOT);
+
+    if (isSlotStillAvailable(slot, slotsPerHistoricalRoot)) {
+      return Optional.empty();
+    }
+    final BeaconState bestState = store.getBlockState(bestBlockRoot);
+    if (bestState == null) {
+      return Optional.empty();
+    }
+
+    int historicalIndex = slot.mod(slotsPerHistoricalRoot).intValue();
+    return Optional.ofNullable(bestState.getBlock_roots().get(historicalIndex));
+  }
+
+  private boolean isSlotStillAvailable(
+      final UnsignedLong slot, final UnsignedLong slotsPerHistoricalRoot) {
+    return bestSlot.compareTo(slotsPerHistoricalRoot) >= 0
```

Comment:
I don't think this check is right - should be something like:
```suggestion
    return bestSlot.compareTo(slot) >= 0 && bestSlot.compareTo(slot.plus(slotsPerHistoricalRoot)) <= 0
```
Also, we need to be careful about subtraction operations on `UnsignedLong`'s.

Reply:
I think this actually did the exact opposite of what the method name suggested, possibly with some other bugs though... In any case I've switched it to use the same check as `get_block_root_at_slot`.